### DISCUSSION TITLE
Question about stats

### DISCUSSION THREAD

**Discussion by markwoon** [umami-software/umami#2544] [General]
I'm trying to pull stats via the API (`/api/websites/site-id/stats`) and have some questions:

1)  According to the docs, `uniques` is the number of unique visitors and `bounces` are the number of returning visitors.  Apart from the fact the `bounces` variable name is very misleading, does that mean that the total number of visitors is `uniques` + `bounces`?
2) How does this concept of visitors relate to the concept of `sessions` from the `/api/websites/site-id/pageviews` API?  I assumed that a visitor can have multiple sessions, but that doesn't seem correct as I'm seeing the number of sessions is always less than the number of `uniques`.
2) Can someone confirm that `totaltime` is the total time spend by everyone during the perioud on the website?  So average time would be `totaltime` / (`uniques` + `bounces`)?
5) What does the `change` number represent?  Change from what?  The previous period?  This does not seem correct based on the numbers I'm seeing.

Any help would be appreciated.

**Comment by Maxime-J** [comment]
It's an interesting topic.

Open source offers transparency but it's a bit opaque on that field outside of code.
I do believe there should be a place where all the mechanics are defined.

In the current version (2.10.0), if I don't make any mistakes:

1/ There's obviously an issue in the docs, `bounces` are not the number of returning visitors.
It's the number of time a visitor has seen only one page of the website in an hour.
_(Just mentioning @mikecao so that the team is aware of that)._

`uniques` is the number of visitors, knowing that it corresponds to the number of unique `sessions`, which leads to your second point.

2/ When someone visits the website, an hash is associated with that visitor, saved as a session id.
A visitor can keep the same session id an all month if its ip or user-agent don't change.

So `sessions` from pageviews API corresponds to the number of unique session ids in the unit period chosen, hence it corresponds to the number of visitors.

3/ `totaltime` is the sum of times between two bounding pageviews per hour and per visitor.
Linking that with the first point, that means bounces are excluded

Average time shown by Umami is `totaltime / (pageviews - bounces)`, corresponding to the average time per pageview.

4/ It's a comparison with the same period of time before the start date of the selected one.
You can see an example in that discussion https://github.com/umami-software/umami/discussions/2497

Hope it helps :)

**Comment by mikecao** [comment]
@Maxime-J is right, our documentation is not correct. Documentation has been updated.

Currently, a session is equivalent to a unique visitor. But our use of the terminology is wrong. In the future we plan to change this so it's more clear.

The new terms will be:
- visitor: a unique user
- pageview: a single view from any visitor
- session: a period of time spent by a visitor (within a certain window, for example 1 hour)
- totaltime: total time spent by a visitor
- bounces: number of times a visitor visits only one page

**Reply by maxiride** [reply]
I hope it is still relevant since more than a year has passed.

> bounces: number of times a visitor visits only one page

As a developer building SPA/landing pages where there is only **one page** like `domain.com`, how can I correctly track bounces? There are no page changes due to the design nature of the site.

From the docs I read:
> Bounce rate
A bounce is defined when a visit has only 1 **event**. A user will enter the website, initially generating a view, then leaving. The bounce rate is calculated by dividing the total number of bounces by unique number of visits.

I've already tried to implement [tracker functions](https://umami.is/docs/tracker-functions) on actions performed on the page to increase the events measured, but that doesn't seem to "raise" the engagement from a metric point of view in umami. Bounces are still measured.

Related #3287 #3518

**Reply by maxiride** [reply]
As confirmed by inspecting the source (if I got it right) only pageview events are considered. So the documentation is outdated/wrong but the comment here reflect the actual implementation.

> bounces: number of times a visitor visits only one page

Wouldn't make sense that a bounce is a visit with only **one** event? Queries could be refactored to count bounces where there are no other events other than the page visit without affecting much of code.

**Comment by markwoon** [comment]
Thanks for the help!  That clears up my questions.

But now I've a few more new ones revolving around the `metrics` endpoint:

1) Is it possible to distinguish between pageviews and unique pageviews?
2) Is there any way to get query string information?  Or is that lost?
3) Is it possible to get more than 100 entries?
4) Even better, can I get metrics based on a partial url?  For example, top 20 pages beginning with `/foo/`?


And for the `events` endpoint:

1) My assumption was that if I provided optional query param `eventName`, it would filter results to only events with that name.  This does not seem to be the case.  Is this user error or bug?
2) How do I pull back the data I sent in with the event?

**Comment by markwoon** [comment]
I see that question 4 above is addressed in issue #2070.

This is a critical API requirement for us...

**Comment by markwoon** [comment]
@Maxime-J @mikecao Any thoughts?

**Comment by mikecao** [comment]
We will be making some updates to the metrics endpoint that should address some of your issues.
1. The new endpoint will return both views and visitors.
2. Query string data is under the query parameters tab.
3. We plan to add paging so you will be able to go through all the records
4. We are adding a search option to filter results.

For events, see https://umami.is/docs/event-data
